File: calculus_and_numerical_methods/finite_difference.py
"""
finite_difference.py
Compute finite difference approximation of the first derivative.
Author: Scott
"""

def finite_difference(f, x, h=1e-5):
    return (f(x + h) - f(x)) / h

if __name__ == "__main__":
    f = lambda x: x**2  # Example function
    x = float(input("Enter x (default 2.0): ") or 2.0)
    print(f"f'(x) ≈ {finite_difference(f, x):.5f}")

File: calculus_and_numerical_methods/finite_difference.cpp

/**
 * Finite Difference Approximation
 * Approximates the derivative of a function using forward difference.
 * Author: Scott Douglass
 * License: MIT
 */

#include <iostream>
#include <cmath>
#include <functional>

double f(double x) {
    return std::sin(x);  // Example function
}

double finite_difference(double x, double h) {
    return (f(x + h) - f(x)) / h;
}

int main() {
    double x = M_PI / 4;
    double h = 1e-5;
    std::cout << "f'(" << x << ") ≈ " << finite_difference(x, h) << std::endl;
    return 0;
}
